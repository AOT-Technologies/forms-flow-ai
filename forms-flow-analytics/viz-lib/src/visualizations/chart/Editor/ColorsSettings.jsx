import React from "react";
import { EditorPropTypes } from "@/visualizations/prop-types";

import PieColorsSettings from "./PieColorsSettings";
import HeatmapColorsSettings from "./HeatmapColorsSettings";
import DefaultColorsSettings from "./DefaultColorsSettings";

const components = {
  pie: PieColorsSettings,
  heatmap: HeatmapColorsSettings,
};

export default function ColorsSettings({ options, ...props }) {
  const Component = components[options.globalSeriesType] || DefaultColorsSettings;
  return <Component options={options} {...props} />;
}

ColorsSettings.propTypes = EditorPropTypes;
