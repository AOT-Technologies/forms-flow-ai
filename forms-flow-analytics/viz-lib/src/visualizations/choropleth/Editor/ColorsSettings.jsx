import React from "react";
import { useDebouncedCallback } from "use-debounce";
import { Section, Select, InputNumber, ColorPicker } from "@/components/visualizations/editor";
import { EditorPropTypes } from "@/visualizations/prop-types";
import ColorPalette from "../ColorPalette";

export default function ColorsSettings({ options, onOptionsChange }) {
  const [onOptionsChangeDebounced] = useDebouncedCallback(onOptionsChange, 200);

  return (
    <React.Fragment>
      <Section>
        <Select
          layout="horizontal"
          label="Clustering mode"
          data-test="Choropleth.Editor.ClusteringMode"
          defaultValue={options.clusteringMode}
          onChange={clusteringMode => onOptionsChange({ clusteringMode })}>
          <Select.Option value="q" data-test="Choropleth.Editor.ClusteringMode.q">
            quantile
          </Select.Option>
          <Select.Option value="e" data-test="Choropleth.Editor.ClusteringMode.e">
            equidistant
          </Select.Option>
          <Select.Option value="k" data-test="Choropleth.Editor.ClusteringMode.k">
            k-means
          </Select.Option>
        </Select>
      </Section>

      <Section>
        <InputNumber
          layout="horizontal"
          label="Steps"
          data-test="Choropleth.Editor.ColorSteps"
          min={3}
          max={11}
          defaultValue={options.steps}
          onChange={steps => onOptionsChangeDebounced({ steps })}
        />
      </Section>

      <Section>
        <ColorPicker
          layout="horizontal"
          label="Min Color"
          interactive
          presetColors={ColorPalette}
          placement="topRight"
          color={options.colors.min}
          triggerProps={{ "data-test": "Choropleth.Editor.Colors.Min" }}
          onChange={min => onOptionsChange({ colors: { min } })}
          addonAfter={<ColorPicker.Label color={options.colors.min} presetColors={ColorPalette} />}
        />
      </Section>

      <Section>
        <ColorPicker
          layout="horizontal"
          label="Max Color"
          interactive
          presetColors={ColorPalette}
          placement="topRight"
          color={options.colors.max}
          triggerProps={{ "data-test": "Choropleth.Editor.Colors.Max" }}
          onChange={max => onOptionsChange({ colors: { max } })}
          addonAfter={<ColorPicker.Label color={options.colors.max} presetColors={ColorPalette} />}
        />
      </Section>

      <Section>
        <ColorPicker
          layout="horizontal"
          label="No value color"
          interactive
          presetColors={ColorPalette}
          placement="topRight"
          color={options.colors.noValue}
          triggerProps={{ "data-test": "Choropleth.Editor.Colors.NoValue" }}
          onChange={noValue => onOptionsChange({ colors: { noValue } })}
          addonAfter={<ColorPicker.Label color={options.colors.noValue} presetColors={ColorPalette} />}
        />
      </Section>

      <Section>
        <ColorPicker
          layout="horizontal"
          label="Background color"
          interactive
          presetColors={ColorPalette}
          placement="topRight"
          color={options.colors.background}
          triggerProps={{ "data-test": "Choropleth.Editor.Colors.Background" }}
          onChange={background => onOptionsChange({ colors: { background } })}
          addonAfter={<ColorPicker.Label color={options.colors.background} presetColors={ColorPalette} />}
        />
      </Section>

      <Section>
        <ColorPicker
          layout="horizontal"
          label="Borders color"
          interactive
          presetColors={ColorPalette}
          placement="topRight"
          color={options.colors.borders}
          triggerProps={{ "data-test": "Choropleth.Editor.Colors.Borders" }}
          onChange={borders => onOptionsChange({ colors: { borders } })}
          addonAfter={<ColorPicker.Label color={options.colors.borders} presetColors={ColorPalette} />}
        />
      </Section>
    </React.Fragment>
  );
}

ColorsSettings.propTypes = EditorPropTypes;
