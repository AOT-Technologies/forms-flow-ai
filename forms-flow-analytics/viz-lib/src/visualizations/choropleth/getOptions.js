import { merge } from "lodash";
import ColorPalette from "./ColorPalette";

const DEFAULT_OPTIONS = {
  mapType: "countries",
  countryCodeColumn: "",
  countryCodeType: "iso_a3",
  valueColumn: "",
  clusteringMode: "e",
  steps: 5,
  valueFormat: "0,0.00",
  noValuePlaceholder: "N/A",
  colors: {
    min: ColorPalette["Light Blue"],
    max: ColorPalette["Dark Blue"],
    background: ColorPalette.White,
    borders: ColorPalette.White,
    noValue: ColorPalette["Light Gray"],
  },
  legend: {
    visible: true,
    position: "bottom-left",
    alignText: "right",
  },
  tooltip: {
    enabled: true,
    template: "<b>{{ @@name }}</b>: {{ @@value }}",
  },
  popup: {
    enabled: true,
    template: "Country: <b>{{ @@name_long }} ({{ @@iso_a2 }})</b>\n<br>\nValue: <b>{{ @@value }}</b>",
  },
};

export default function getOptions(options) {
  return merge({}, DEFAULT_OPTIONS, options);
}
