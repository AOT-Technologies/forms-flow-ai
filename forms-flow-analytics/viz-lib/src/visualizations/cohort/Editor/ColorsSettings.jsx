import { isFinite } from "lodash";
import React from "react";
import { Section, ColorPicker, InputNumber } from "@/components/visualizations/editor";
import { EditorPropTypes } from "@/visualizations/prop-types";
import DefaultColorPalette from "@/visualizations/ColorPalette";

const ColorPalette = {
  White: "#FFFFFF",
  ...DefaultColorPalette,
};

const minSteps = 3;
const maxSteps = 20;

function validateSteps(value) {
  value = isFinite(value) ? value : parseInt(value, 10);
  value = isFinite(value) ? value : 0;
  return Math.max(minSteps, Math.min(value, maxSteps));
}

export default function ColorsSettings({ options, onOptionsChange }) {
  return (
    <React.Fragment>
      <Section>
        <ColorPicker
          layout="horizontal"
          label="Min Color"
          presetColors={ColorPalette}
          interactive
          color={options.colors.min}
          onChange={min => onOptionsChange({ colors: { min } })}
          addonAfter={<ColorPicker.Label color={options.colors.min} presetColors={ColorPalette} />}
        />
      </Section>
      <Section>
        <ColorPicker
          layout="horizontal"
          label="Max Color"
          presetColors={ColorPalette}
          interactive
          color={options.colors.max}
          onChange={max => onOptionsChange({ colors: { max } })}
          addonAfter={<ColorPicker.Label color={options.colors.max} presetColors={ColorPalette} />}
        />
      </Section>
      <Section>
        <InputNumber
          layout="horizontal"
          label="Steps"
          min={minSteps}
          max={maxSteps}
          value={options.colors.steps}
          onChange={value => onOptionsChange({ colors: { steps: validateSteps(value) } })}
        />
      </Section>
    </React.Fragment>
  );
}

ColorsSettings.propTypes = EditorPropTypes;
