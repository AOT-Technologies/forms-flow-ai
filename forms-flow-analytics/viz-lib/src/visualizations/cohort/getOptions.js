import { merge } from "lodash";
import ColorPalette from "@/visualizations/ColorPalette";

const DEFAULT_OPTIONS = {
  timeInterval: "daily",
  mode: "diagonal",
  dateColumn: "date",
  stageColumn: "day_number",
  totalColumn: "total",
  valueColumn: "value",

  showTooltips: true,
  percentValues: true,

  timeColumnTitle: "Time",
  peopleColumnTitle: "Users",
  stageColumnTitle: "{{ @ }}",

  numberFormat: "0,0[.]00",
  percentFormat: "0.00%",
  noValuePlaceholder: "-",

  colors: {
    min: "#ffffff",
    max: ColorPalette["Dark Blue"],
    steps: 7,
  },
};

export default function getOptions(options) {
  return merge({}, DEFAULT_OPTIONS, options);
}
