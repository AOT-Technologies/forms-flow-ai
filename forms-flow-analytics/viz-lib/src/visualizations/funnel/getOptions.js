import { isFinite, map, merge, includes } from "lodash";

const DEFAULT_OPTIONS = {
  stepCol: { colName: null, displayAs: "Steps" },
  valueCol: { colName: null, displayAs: "Value" },
  autoSort: true,
  sortKeyCol: { colName: null, reverse: false },
  itemsLimit: 100,
  percentValuesRange: { min: 0.01, max: 1000.0 },
  numberFormat: "0,0[.]00",
  percentFormat: "0[.]00%",
};

export default function getOptions(options, { columns }) {
  options = merge({}, DEFAULT_OPTIONS, options);

  // Validate
  const availableColumns = map(columns, c => c.name);
  if (!includes(availableColumns, options.stepCol.colName)) {
    options.stepCol.colName = null;
  }
  if (!includes(availableColumns, options.valueCol.colName)) {
    options.valueCol.colName = null;
  }
  if (!includes(availableColumns, options.sortKeyCol.colName)) {
    options.sortKeyCol.colName = null;
  }

  if (!isFinite(options.itemsLimit)) {
    options.itemsLimit = DEFAULT_OPTIONS.itemsLimit;
  }
  if (options.itemsLimit < 2) {
    options.itemsLimit = 2;
  }

  if (options.autoSort) {
    options.sortKeyCol.colName = options.valueCol.colName;
    options.sortKeyCol.reverse = true;
  }

  return options;
}
