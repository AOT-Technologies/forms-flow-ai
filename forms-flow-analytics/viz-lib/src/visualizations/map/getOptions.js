import { merge } from "lodash";

const DEFAULT_OPTIONS = {
  latColName: "lat",
  lonColName: "lon",
  classify: null,
  groups: {},
  mapTileUrl: "//{s}.tile.openstreetmap.org/{z}/{x}/{y}.png",
  clusterMarkers: true,
  customizeMarkers: false,
  iconShape: "marker",
  iconFont: "circle",
  foregroundColor: "#ffffff",
  backgroundColor: "#356AFF",
  borderColor: "#356AFF",
  bounds: null,
  tooltip: {
    enabled: false,
    template: "",
  },
  popup: {
    enabled: true,
    template: "",
  },
};

export default function getOptions(options) {
  options = merge({}, DEFAULT_OPTIONS, options);
  options.mapTileUrl = options.mapTileUrl || DEFAULT_OPTIONS.mapTileUrl;

  // Backward compatibility
  if (options.classify === "none") {
    options.classify = null;
  }

  return options;
}
