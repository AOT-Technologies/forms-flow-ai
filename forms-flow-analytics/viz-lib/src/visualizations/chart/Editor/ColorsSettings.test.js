import { after } from "lodash";
import React from "react";
import enzyme from "enzyme";

import getOptions from "../getOptions";
import ColorsSettings from "./ColorsSettings";

function findByTestID(wrapper, testId) {
  return wrapper.find(`[data-test="${testId}"]`);
}

function mount(options, done) {
  options = getOptions(options);
  return enzyme.mount(
    <ColorsSettings
      visualizationName="Test"
      data={{
        columns: [
          { name: "a", type: "string" },
          { name: "b", type: "number" },
        ],
        rows: [{ a: "v", b: 3.14 }],
      }}
      options={options}
      onOptionsChange={changedOptions => {
        expect(changedOptions).toMatchSnapshot();
        done();
      }}
    />
  );
}

describe("Visualizations -> Chart -> Editor -> Colors Settings", () => {
  describe("for pie", () => {
    test("Changes series color", done => {
      const el = mount(
        {
          globalSeriesType: "pie",
          columnMapping: { a: "x", b: "y" },
        },
        done
      );

      findByTestID(el, "Chart.Series.v.Color")
        .find(".color-picker-trigger")
        .last()
        .simulate("click");
      findByTestID(el, "ColorPicker")
        .last()
        .find("input")
        .simulate("change", { target: { value: "red" } });
    });
  });

  describe("for heatmap", () => {
    test("Changes color scheme", done => {
      const el = mount(
        {
          globalSeriesType: "heatmap",
          columnMapping: { a: "x", b: "y" },
        },
        done
      );

      findByTestID(el, "Chart.Colors.Heatmap.ColorScheme")
        .last()
        .simulate("click");
      findByTestID(el, "Chart.Colors.Heatmap.ColorScheme.RdBu")
        .last()
        .simulate("click");
    });

    test("Sets custom color scheme", done => {
      const el = mount(
        {
          globalSeriesType: "heatmap",
          columnMapping: { a: "x", b: "y" },
          colorScheme: "Custom...",
        },
        after(2, done)
      ); // we will perform 2 actions, so call `done` after all of them completed

      findByTestID(el, "Chart.Colors.Heatmap.MinColor")
        .find(".color-picker-trigger")
        .last()
        .simulate("click");
      findByTestID(el, "ColorPicker")
        .last()
        .find("input")
        .simulate("change", { target: { value: "yellow" } });

      findByTestID(el, "Chart.Colors.Heatmap.MaxColor")
        .find(".color-picker-trigger")
        .last()
        .simulate("click");
      findByTestID(el, "ColorPicker")
        .last()
        .find("input")
        .simulate("change", { target: { value: "red" } });
    });
  });

  describe("for all except of pie and heatmap", () => {
    test("Changes series color", done => {
      const el = mount(
        {
          globalSeriesType: "column",
          columnMapping: { a: "x", b: "y" },
        },
        done
      );

      findByTestID(el, "Chart.Series.b.Color")
        .find(".color-picker-trigger")
        .last()
        .simulate("click");

      findByTestID(el, "ColorPicker")
        .last()
        .find("input")
        .simulate("change", { target: { value: "red" } });
    });
  });
});
