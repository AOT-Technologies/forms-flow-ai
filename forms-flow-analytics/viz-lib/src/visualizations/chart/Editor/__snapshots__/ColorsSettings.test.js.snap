// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Visualizations -> Chart -> Editor -> Colors Settings for all except of pie and heatmap Changes series color 1`] = `
Object {
  "seriesOptions": Object {
    "b": Object {
      "color": "#FF0000",
    },
  },
}
`;

exports[`Visualizations -> Chart -> Editor -> Colors Settings for heatmap Changes color scheme 1`] = `
Object {
  "colorScheme": "RdBu",
}
`;

exports[`Visualizations -> Chart -> Editor -> Colors Settings for heatmap Sets custom color scheme 1`] = `
Object {
  "heatMinColor": "#FFFF00",
}
`;

exports[`Visualizations -> Chart -> Editor -> Colors Settings for heatmap Sets custom color scheme 2`] = `
Object {
  "heatMaxColor": "#FF0000",
}
`;

exports[`Visualizations -> Chart -> Editor -> Colors Settings for pie Changes series color 1`] = `
Object {
  "valuesOptions": Object {
    "v": Object {
      "color": "#FF0000",
    },
  },
}
`;
